[FSDP] Implement sharded_state_dict and load_sharded_state_dict

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77356

Implement ShardedTensor compatible sharded_state_dict() and load_sharded_state_dict().

Algorithm overview:
sharded_state_dict():
  1. Call summon_full_parameters().
  2. For each unflattened, non-sharded parameter.
      2.1 Call chunk() to get the local shard of the parameter.
      2.2 Create a ShardedTensor.
  3. Replace the tensor in the state_dict with the newly created ShardedTensor.

load_sharded_state_dict():
   1. For each unflattened, sharded parameter (ShardedTensor) in the given state_dict:
       1.1 Pop out from the state_dict.
       1.2 Do allgather to reconstruct the unflattened, non-sharded parameter.
   2. Create a FlatParameter with the unflattened, non-sharded parameters.
   3. Shard the newly created FlatParameter.
   4. Insert the new FlatParameter into the state_dict.

Differential Revision: [D36284983](https://our.internmc.facebook.com/intern/diff/D36284983/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D36284983/)!

fegin committed:
[FSDP] Implement sharded_state_dict and load_sharded_state_dict

Implement ShardedTensor compatible sharded_state_dict() and load_sharded_state_dict().

Algorithm overview:
sharded_state_dict():
  1. Call summon_full_parameters().
  2. For each unflattened, non-sharded parameter.
      2.1 Call chunk() to get the local shard of the parameter.
      2.2 Create a ShardedTensor.
  3. Replace the tensor in the state_dict with the newly created ShardedTensor.

load_sharded_state_dict():
   1. For each unflattened, sharded parameter (ShardedTensor) in the given state_dict:
       1.1 Pop out from the state_dict.
       1.2 Do allgather to reconstruct the unflattened, non-sharded parameter.
   2. Create a FlatParameter with the unflattened, non-sharded parameters.
   3. Shard the newly created FlatParameter.
   4. Insert the new FlatParameter into the state_dict.

Differential Revision: [D36284983](https://our.internmc.facebook.com/intern/diff/D36284983/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D36284983/)!

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_state_dict and load_sharded_state_dict"


Implement ShardedTensor compatible sharded_state_dict() and load_sharded_state_dict().

Algorithm overview:
sharded_state_dict():
  1. Call summon_full_parameters().
  2. For each unflattened, non-sharded parameter.
      2.1 Call chunk() to get the local shard of the parameter.
      2.2 Create a ShardedTensor.
  3. Replace the tensor in the state_dict with the newly created ShardedTensor.

load_sharded_state_dict():
   1. For each unflattened, sharded parameter (ShardedTensor) in the given state_dict:
       1.1 Pop out from the state_dict.
       1.2 Do allgather to reconstruct the unflattened, non-sharded parameter.
   2. Create a FlatParameter with the unflattened, non-sharded parameters.
   3. Shard the newly created FlatParameter.
   4. Insert the new FlatParameter into the state_dict.

Differential Revision: [D36284983](https://our.internmc.facebook.com/intern/diff/D36284983/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D36284983/)!

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_state_dict and load_sharded_state_dict"


Implement ShardedTensor compatible sharded_state_dict() and load_sharded_state_dict().

Algorithm overview:
sharded_state_dict():
  1. Call summon_full_parameters().
  2. For each unflattened, non-sharded parameter.
      2.1 Call chunk() to get the local shard of the parameter.
      2.2 Create a ShardedTensor.
  3. Replace the tensor in the state_dict with the newly created ShardedTensor.

load_sharded_state_dict():
   1. For each unflattened, sharded parameter (ShardedTensor) in the given state_dict:
       1.1 Pop out from the state_dict.
       1.2 Do allgather to reconstruct the unflattened, non-sharded parameter.
   2. Create a FlatParameter with the unflattened, non-sharded parameters.
   3. Shard the newly created FlatParameter.
   4. Insert the new FlatParameter into the state_dict.

Differential Revision: [D36284983](https://our.internmc.facebook.com/intern/diff/D36284983/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D36284983/)!

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_state_dict and load_sharded_state_dict"


Implement ShardedTensor compatible sharded_state_dict() and load_sharded_state_dict().

Algorithm overview:
sharded_state_dict():
  1. Call summon_full_parameters().
  2. For each unflattened, non-sharded parameter.
      2.1 Call chunk() to get the local shard of the parameter.
      2.2 Create a ShardedTensor.
  3. Replace the tensor in the state_dict with the newly created ShardedTensor.

load_sharded_state_dict():
   1. For each unflattened, sharded parameter (ShardedTensor) in the given state_dict:
       1.1 Pop out from the state_dict.
       1.2 Do allgather to reconstruct the unflattened, non-sharded parameter.
   2. Create a FlatParameter with the unflattened, non-sharded parameters.
   3. Shard the newly created FlatParameter.
   4. Insert the new FlatParameter into the state_dict.

Differential Revision: [D36284983](https://our.internmc.facebook.com/intern/diff/D36284983/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D36284983/)!

[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1830,14 +1873,72 @@ def _local_pre_load_state_dict_hook(
+        # TODO: Reduce the communication by using only one _all_gather_base to
+        # gather all the parameters in this layer. This can be achieved by
+        # concatenated all the local shards and then append the padding.
+        for module_name, _, param_name in self.module.flat_param._param_infos:
```
could we create a Github issue regarding followup optimizations as we discussed offline, like coalesced all gather to reduce communication here?

zhaojuanmao reviewed (approved):
looks good! some mypy is messed up

fegin committed:
Update on "[FSDP] Implement sharded_state_dict and load_sharded_state_dict"


Implement ShardedTensor compatible sharded_state_dict() and load_sharded_state_dict().

Algorithm overview:
sharded_state_dict():
  1. Call summon_full_parameters().
  2. For each unflattened, non-sharded parameter.
      2.1 Call chunk() to get the local shard of the parameter.
      2.2 Create a ShardedTensor.
  3. Replace the tensor in the state_dict with the newly created ShardedTensor.

load_sharded_state_dict():
   1. For each unflattened, sharded parameter (ShardedTensor) in the given state_dict:
       1.1 Pop out from the state_dict.
       1.2 Do allgather to reconstruct the unflattened, non-sharded parameter.
   2. Create a FlatParameter with the unflattened, non-sharded parameters.
   3. Shard the newly created FlatParameter.
   4. Insert the new FlatParameter into the state_dict.

Differential Revision: [D36284983](https://our.internmc.facebook.com/intern/diff/D36284983/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D36284983/)!

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_state_dict and load_sharded_state_dict"


Implement ShardedTensor compatible sharded_state_dict() and load_sharded_state_dict().

Algorithm overview:
sharded_state_dict():
  1. Call summon_full_parameters().
  2. For each unflattened, non-sharded parameter.
      2.1 Call chunk() to get the local shard of the parameter.
      2.2 Create a ShardedTensor.
  3. Replace the tensor in the state_dict with the newly created ShardedTensor.

load_sharded_state_dict():
   1. For each unflattened, sharded parameter (ShardedTensor) in the given state_dict:
       1.1 Pop out from the state_dict.
       1.2 Do allgather to reconstruct the unflattened, non-sharded parameter.
   2. Create a FlatParameter with the unflattened, non-sharded parameters.
   3. Shard the newly created FlatParameter.
   4. Insert the new FlatParameter into the state_dict.

Differential Revision: [D36284983](https://our.internmc.facebook.com/intern/diff/D36284983/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D36284983/)!

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_state_dict and load_sharded_state_dict"


Implement ShardedTensor compatible sharded_state_dict() and load_sharded_state_dict().

Algorithm overview:
sharded_state_dict():
  1. Call summon_full_parameters().
  2. For each unflattened, non-sharded parameter.
      2.1 Call chunk() to get the local shard of the parameter.
      2.2 Create a ShardedTensor.
  3. Replace the tensor in the state_dict with the newly created ShardedTensor.

load_sharded_state_dict():
   1. For each unflattened, sharded parameter (ShardedTensor) in the given state_dict:
       1.1 Pop out from the state_dict.
       1.2 Do allgather to reconstruct the unflattened, non-sharded parameter.
   2. Create a FlatParameter with the unflattened, non-sharded parameters.
   3. Shard the newly created FlatParameter.
   4. Insert the new FlatParameter into the state_dict.

Differential Revision: [D36284983](https://our.internmc.facebook.com/intern/diff/D36284983/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D36284983/)!

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_state_dict and load_sharded_state_dict"


Implement ShardedTensor compatible sharded_state_dict() and load_sharded_state_dict().

Algorithm overview:
sharded_state_dict():
  1. Call summon_full_parameters().
  2. For each unflattened, non-sharded parameter.
      2.1 Call chunk() to get the local shard of the parameter.
      2.2 Create a ShardedTensor.
  3. Replace the tensor in the state_dict with the newly created ShardedTensor.

load_sharded_state_dict():
   1. For each unflattened, sharded parameter (ShardedTensor) in the given state_dict:
       1.1 Pop out from the state_dict.
       1.2 Do allgather to reconstruct the unflattened, non-sharded parameter.
   2. Create a FlatParameter with the unflattened, non-sharded parameters.
   3. Shard the newly created FlatParameter.
   4. Insert the new FlatParameter into the state_dict.

Differential Revision: [D36284983](https://our.internmc.facebook.com/intern/diff/D36284983/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D36284983/)!

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_state_dict and load_sharded_state_dict"


Implement ShardedTensor compatible sharded_state_dict() and load_sharded_state_dict().

Algorithm overview:
sharded_state_dict():
  1. Call summon_full_parameters().
  2. For each unflattened, non-sharded parameter.
      2.1 Call chunk() to get the local shard of the parameter.
      2.2 Create a ShardedTensor.
  3. Replace the tensor in the state_dict with the newly created ShardedTensor.

load_sharded_state_dict():
   1. For each unflattened, sharded parameter (ShardedTensor) in the given state_dict:
       1.1 Pop out from the state_dict.
       1.2 Do allgather to reconstruct the unflattened, non-sharded parameter.
   2. Create a FlatParameter with the unflattened, non-sharded parameters.
   3. Shard the newly created FlatParameter.
   4. Insert the new FlatParameter into the state_dict.

Differential Revision: [D36284983](https://our.internmc.facebook.com/intern/diff/D36284983/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D36284983/)!

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_state_dict and load_sharded_state_dict"


Implement ShardedTensor compatible sharded_state_dict() and load_sharded_state_dict().

Algorithm overview:
sharded_state_dict():
  1. Call summon_full_parameters().
  2. For each unflattened, non-sharded parameter.
      2.1 Call chunk() to get the local shard of the parameter.
      2.2 Create a ShardedTensor.
  3. Replace the tensor in the state_dict with the newly created ShardedTensor.

load_sharded_state_dict():
   1. For each unflattened, sharded parameter (ShardedTensor) in the given state_dict:
       1.1 Pop out from the state_dict.
       1.2 Do allgather to reconstruct the unflattened, non-sharded parameter.
   2. Create a FlatParameter with the unflattened, non-sharded parameters.
   3. Shard the newly created FlatParameter.
   4. Insert the new FlatParameter into the state_dict.

Differential Revision: [D36284983](https://our.internmc.facebook.com/intern/diff/D36284983/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D36284983/)!

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_state_dict and load_sharded_state_dict"


Implement ShardedTensor compatible sharded_state_dict() and load_sharded_state_dict().

Algorithm overview:
sharded_state_dict():
  1. Call summon_full_parameters().
  2. For each unflattened, non-sharded parameter.
      2.1 Call chunk() to get the local shard of the parameter.
      2.2 Create a ShardedTensor.
  3. Replace the tensor in the state_dict with the newly created ShardedTensor.

load_sharded_state_dict():
   1. For each unflattened, sharded parameter (ShardedTensor) in the given state_dict:
       1.1 Pop out from the state_dict.
       1.2 Do allgather to reconstruct the unflattened, non-sharded parameter.
   2. Create a FlatParameter with the unflattened, non-sharded parameters.
   3. Shard the newly created FlatParameter.
   4. Insert the new FlatParameter into the state_dict.

Differential Revision: [D36284983](https://our.internmc.facebook.com/intern/diff/D36284983/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D36284983/)!

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_state_dict and load_sharded_state_dict"


Implement ShardedTensor compatible sharded_state_dict() and load_sharded_state_dict().

Algorithm overview:
sharded_state_dict():
  1. Call summon_full_parameters().
  2. For each unflattened, non-sharded parameter.
      2.1 Call chunk() to get the local shard of the parameter.
      2.2 Create a ShardedTensor.
  3. Replace the tensor in the state_dict with the newly created ShardedTensor.

load_sharded_state_dict():
   1. For each unflattened, sharded parameter (ShardedTensor) in the given state_dict:
       1.1 Pop out from the state_dict.
       1.2 Do allgather to reconstruct the unflattened, non-sharded parameter.
   2. Create a FlatParameter with the unflattened, non-sharded parameters.
   3. Shard the newly created FlatParameter.
   4. Insert the new FlatParameter into the state_dict.

Differential Revision: [D36284983](https://our.internmc.facebook.com/intern/diff/D36284983/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D36284983/)!

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_state_dict and load_sharded_state_dict"


Implement ShardedTensor compatible sharded_state_dict() and load_sharded_state_dict().

Algorithm overview:
sharded_state_dict():
  1. Call summon_full_parameters().
  2. For each unflattened, non-sharded parameter.
      2.1 Call chunk() to get the local shard of the parameter.
      2.2 Create a ShardedTensor.
  3. Replace the tensor in the state_dict with the newly created ShardedTensor.

load_sharded_state_dict():
   1. For each unflattened, sharded parameter (ShardedTensor) in the given state_dict:
       1.1 Pop out from the state_dict.
       1.2 Do allgather to reconstruct the unflattened, non-sharded parameter.
   2. Create a FlatParameter with the unflattened, non-sharded parameters.
   3. Shard the newly created FlatParameter.
   4. Insert the new FlatParameter into the state_dict.

Differential Revision: [D36284983](https://our.internmc.facebook.com/intern/diff/D36284983/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D36284983/)!

[ghstack-poisoned]
